File: kraken/src/main/java/glitch/kraken/services/request/TeamsRequest.java
package glitch.kraken.services.request;

import glitch.api.http.HttpClient;
import glitch.api.http.Routes;
import glitch.kraken.object.json.Team;
import glitch.kraken.object.json.collections.Teams;
import glitch.service.AbstractRequest;
import reactor.core.publisher.Mono;

public class TeamsRequest extends AbstractRequest<Team, Teams> {
    private Integer limit;
    private Integer offset;

    public TeamsRequest(HttpClient httpClient) {
        super(httpClient, Routes.get("/teams").newRequest());
    }

    public TeamsRequest setLimit(Integer limit) {
        this.limit = limit;
        return this;
    }

    public TeamsRequest setOffset(Integer offset) {
        this.offset = offset;
        return this;
    }

    @Override
    public Mono<Teams> get() {

        if (limit != null && limit > 0 && limit <= 100) {
            request.queryParam("limit", limit);
        }

        if (offset != null && offset >= 0) {
            request.queryParam("offset", offset);
        }

        return httpClient.exchangeAs(request, Teams.class);
    }
}

File: kraken/src/main/kotlin/io/glitchlib/v5/service/request/TeamsRequest.kt
package io.glitchlib.v5.service.request

import io.glitchlib.internal.http.AbstractRequest
import io.glitchlib.internal.http.HttpRequest

class TeamsRequest internal constructor() : AbstractRequest {
    //    AbstractRequest<Team, Teams>(httpClient, Routes.Companion.get("/teams").newRequest()) {
    private var limit: Int? = null
    private var offset: Int? = null

    fun limit(limit: Int) = apply {
        this.limit = limit
    }

    fun offset(offset: Int) = apply {
        this.offset = offset
    }

    override fun invoke(): HttpRequest.() -> Unit = {
        if (limit != null && limit!! > 0 && limit!! <= 100) {
            addQueryParameters("limit", limit!!.toString())
        }

        if (offset != null && offset!! >= 0) {
            addQueryParameters("offset", offset!!.toString())
        }

    }
}
